File: backend/app/supabase_auth.py
import jwt
from jwt import PyJWKClient

from app.config import settings

_jwks: PyJWKClient | None = None


def _jwks_client() -> PyJWKClient:
    global _jwks
    if _jwks is None:
        _jwks = PyJWKClient(settings.supabase_url.rstrip("/") + "/auth/v1/.well-known/jwks.json")
    return _jwks


def decode_supabase_token(token: str) -> dict:
    header = jwt.get_unverified_header(token)
    algorithm = header.get("alg")
    options = {"verify_aud": False}

    if algorithm == "HS256":
        return jwt.decode(
            token,
            settings.supabase_jwt_secret,
            algorithms=["HS256"],
            options=options,
        )

    signing_key = _jwks_client().get_signing_key_from_jwt(token)
    return jwt.decode(
        token,
        signing_key.key,
        algorithms=[algorithm] if algorithm else None,
        options=options,
    )


def get_token_subject(token: str) -> str | None:
    try:
        return decode_supabase_token(token).get("sub")
    except jwt.PyJWTError:
        return None

File: backend2/src/supabaseAuth.ts
import { createRemoteJWKSet, decodeProtectedHeader, importJWK, jwtVerify } from "jose";

import { settings } from "./config.js";

const jwks = createRemoteJWKSet(new URL(`${settings.supabaseUrl.replace(/\/+$/, "")}/auth/v1/.well-known/jwks.json`));

export async function decodeSupabaseToken(token: string): Promise<Record<string, unknown>> {
  const header = decodeProtectedHeader(token);
  if (header.alg === "HS256") {
    const key = await importJWK(
      { kty: "oct", k: Buffer.from(settings.supabaseJwtSecret, "utf8").toString("base64url"), alg: "HS256" },
      "HS256"
    );
    const verified = await jwtVerify(token, key, { algorithms: ["HS256"] });
    return verified.payload as Record<string, unknown>;
  }
  const verified = await jwtVerify(token, jwks);
  return verified.payload as Record<string, unknown>;
}

export async function getTokenSubject(token: string): Promise<string | null> {
  try {
    const payload = await decodeSupabaseToken(token);
    return typeof payload.sub === "string" ? payload.sub : null;
  } catch {
    return null;
  }
}
